File: leetcode/graph/305_Number_of_Islands_II/codes/union.py
from typing import List

class Solution:
    def numIslands2(self, m: int, n: int, positions: List[List[int]]) -> List[int]:
        parent = {}
        rank = {}
        count = 0
        res = []

        def find(x):
            if parent[x] != x:
                parent[x] = find(parent[x])
            return parent[x]

        def union(x, y):
            rootX, rootY = find(x), find(y)
            if rootX == rootY:
                return False
            if rank[rootX] > rank[rootY]:
                parent[rootY] = rootX
            elif rank[rootX] < rank[rootY]:
                parent[rootX] = rootY
            else:
                parent[rootY] = rootX
                rank[rootX] += 1
            return True

        for i, j in positions:
            pos = (i, j)
            if pos in parent:
                res.append(count)
                continue
            parent[pos] = pos
            rank[pos] = 0
            count += 1
            for dx, dy in [(-1,0),(1,0),(0,-1),(0,1)]:
                ni, nj = i + dx, j + dy
                npos = (ni, nj)
                if npos in parent:
                    if union(pos, npos):
                        count -= 1
            res.append(count)
        return res

File: leetcode/graph/305_Number_of_Islands_II/codes/union.c
#define MAX 10000

int find(int* parent, int x) {
    if (parent[x] != x)
        parent[x] = find(parent, parent[x]);
    return parent[x];
}

int union_set(int* parent, int* rank, int x, int y) {
    int xr = find(parent, x);
    int yr = find(parent, y);
    if (xr == yr) return 0;
    if (rank[xr] < rank[yr]) parent[xr] = yr;
    else if (rank[xr] > rank[yr]) parent[yr] = xr;
    else { parent[yr] = xr; rank[xr]++; }
    return 1;
}

int* numIslands2(int m, int n, int** positions, int positionsSize, int* positionsColSize, int* returnSize) {
    int* res = malloc(sizeof(int) * positionsSize);
    int parent[MAX], rank[MAX];
    for (int i = 0; i < m * n; i++) parent[i] = -1, rank[i] = 0;
    int count = 0;

    int dirs[4][2] = {{0,1},{1,0},{0,-1},{-1,0}};

    for (int k = 0; k < positionsSize; k++) {
        int r = positions[k][0], c = positions[k][1], idx = r * n + c;
        if (parent[idx] != -1) {
            res[k] = count;
            continue;
        }
        parent[idx] = idx;
        count++;

        for (int d = 0; d < 4; d++) {
            int nr = r + dirs[d][0], nc = c + dirs[d][1];
            int nidx = nr * n + nc;
            if (nr >= 0 && nr < m && nc >= 0 && nc < n && parent[nidx] != -1) {
                if (union_set(parent, rank, idx, nidx)) count--;
            }
        }
        res[k] = count;
    }

    *returnSize = positionsSize;
    return res;
}
